### PR TITLE
update settings for `pydantic` 2

### LINKED ISSUE #15290: Simplify PrefectBaseSettings
### Describe the current behavior

Prefect's `settings.py` is roughly ~150 custom `Setting` objects which is dynamically collected into a `BaseSettings` object at runtime. By dynamically collecting these at runtime, this 

- obfuscates typing _and_ 
- makes it difficult to ergonomically inherit from in derivative frameworks (like ControlFlow). 

For instance, it's difficult to change the default value of a single setting given it's current design, but could be trivial if `PrefectBaseSetting` was a simple static subclass of Pydantic's `BaseSetting`. 

From examining it's original PRs, the current design was taken on:

- to attach validators to fields themselves, and not as field_validators on the Settings object. 
- to create `settings dependent on the value of other settings`.

These two outcomes required a sophisticated design two years ago, but are now first-class concepts in Pydantic: `AfterValidator` and `computed_field`, respectively. 

### Describe the proposed behavior

I propose having Prefect's BaseSetting's object be a canonical subclass of Pydantic's BaseSetting:

```python
class PrefectBaseSetting(pydantic_settings.BaseSetting)
    PREFECT_WHATEVER_1 = Field(...)
    ...
    PREFECT_WHATEVER_150 = Field(...)
```

### Example Use

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by zzstoatzz**
> [!NOTE]
> I'd like to suggest we cut an `rc` (`3.1.0`) after merging this PR, since so many low level changes were required. Most complexity here comes from updating the tests, but out of abundance of caution I think an `rc` would be safe.

closes #15290, which well articulates the motivation. tl;dr a lot of the existing custom behavior now comes for free with first class support in pydantic 2 by using `BaseSettings` in a more standard manner

in this PR:
- refactored `Settings` class to use `BaseSettings` with `Field` definitions
  - this involves moving the callback logic to a model after validator as described [here](https://github.com/pydantic/pydantic/issues/9789)
  - I intentionally didn't nest settings objects but suggest a straw man nesting via comments so we can discuss
- simplified `Setting` implementation that wraps `Settings` attrs (kept to avoid broader changes here)
- we preserve `from prefect.settings import PREEFCT_{SOMETHING}` (also keeps internal use for now)
  - using `__getattr__` in this way requires us to define all settings in `__all__` that we want discoverable by LSPs
- update settings and profiles utils to use the new settings object
  - `to_environment_variables`, `copy_with_update` etc
- remove unused `experimental` code (open to reintroducing this if we think it appropriate)

### TODOs
- [x] Remove unused `experimental` code
- [x] Move all settings under a `class Settings(BaseSettings)`
- [x] Move callback logic to validators
- [x] Preserve access to `Setting` objects like `PREFECT_API_KEY`
- [x] Refactor settings utils to work with updated settings
- [x] Refactor profile utils to work with updated settings
- [x] Update profile tests
- [x] Update settings tests
- [x] Solve issues with setting fixtures observed using `xdist`

___
as mentioned by @aaazzam, we can recreate the order-dependent templating something like

```python
PREFECT_MEMO_STORE_PATH: Annotated[
    Path,
    AfterValidator(lambda v: v.expanduser()),
    AfterValidator(_hydrate_with_settings(["PREFECT_HOME"])),
] = Field(
    default=Path("${PREFECT_HOME}") / "memo_store.toml",
    description="""The path to the memo store file.""",
)
```
but I thought we might save these for follow ups so as not to expand the surface area here

### CODE REVIEW

File: src/prefect/server/api/admin.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ async def read_settings() -> prefect.settings.Settings:
 
     Secret setting values will be obfuscated.
     """
-    return prefect.settings.get_current_settings().with_obfuscated_secrets()
+    return prefect.settings.get_current_settings()
```
Comment (zzstoatzz) [inline src/prefect/server/api/admin.py:23]: secrets are obfuscated by default unless `include_secrets=True` is passed as `context`

File: src/prefect/settings.py [hunk 2]
```diff
@@ -1793,101 +1412,109 @@ def copy_with_update(
...
+# Settings utils
 
 # Functions to instantiate `Settings` instances
 
-_DEFAULTS_CACHE: Optional[Settings] = None
-_FROM_ENV_CACHE: Dict[int, Settings] = {}
+
+def _cast_settings(
+    settings: Union[Dict[Union[str, Setting], Any], Any],
+) -> Dict[Setting, Any]:
```
Comment (desertaxle) [inline src/prefect/settings.py:1498]: Can you add a doc string to this function? It's not immediately clear what it's used for.

**Review by desertaxle — APPROVED** [review]

File: src/prefect/logging/configuration.py [hunk 3]
```diff
@@ -31,18 +31,24 @@ def load_logging_config(path: Path) -> dict:
...
+
+    # Build a mapping from environment variable names to values
+    mapping = {}
+    for field_name, value in current_settings.model_dump().items():
+        if value is not None:
+            env_var_name = (
+                current_settings.model_config.get("env_prefix", "PREFECT_")
+                + field_name.upper()
+            )
+            mapping[env_var_name] = str(value)
```
[the remark this answers is not in the document]
Comment (zzstoatzz) [reply src/prefect/logging/configuration.py]: good call, we do already have `to_environment_variables` that does the same thing here. updated in [fdbc373](https://github.com/PrefectHQ/prefect/pull/15412/commits/fdbc3731ab69dddd2163a7f133c3f0f623eb4b31)

File: src/prefect/settings.py [hunk 4]
```diff
@@ -1,59 +1,27 @@
...
-
-Generally, settings should be accessed with `SETTING_VARIABLE.value()` which will
-pull the current `Settings` instance from the current `SettingsContext` and retrieve
-the value of the relevant setting.
-
-Accessing a setting's value will also call the `Setting.value_callback` which allows
-settings to be dynamically modified on retrieval. This allows us to make settings
-dependent on the value of other settings or perform other dynamic effects.
+The recommended user-facing way to access Prefect settings is to use the `get_current_settings()` function, as it will
+check any existing `SettingsContext` for a settings instance or automatically create one if not found.
```
[the remark this answers is not in the document]
Comment (zzstoatzz) [reply src/prefect/settings.py]: that's a good point - updated in [623b3a4](https://github.com/PrefectHQ/prefect/pull/15412/commits/623b3a44911380c639d13e11bb093a80a002c1fd)

File: src/prefect/settings.py (hunk 2 above)
Comment (zzstoatzz) [reply src/prefect/settings.py:1498]: yep, added in [c56164d](https://github.com/PrefectHQ/prefect/pull/15412/commits/c56164dc2f0f3aa0680c7138b1f37c8d03376c33)

File: src/prefect/settings.py [hunk 5]
```diff
@@ -408,1388 +209,1015 @@ def warn_on_misconfigured_api_url(values):
...
+        description="The log level of the runner's webserver.",
+    )
+
+    deployment_concurrency_slot_wait_seconds: Annotated[float, Field(ge=0)] = Field(
+        default=30.0,
+        description=(
+            "The number of seconds to wait before retrying when a deployment flow run"
+            " cannot secure a concurrency slot from the server."
+        ),
+    )
```
Comment (zzstoatzz) [inline src/prefect/settings.py:1225]: here is the new setting added from https://github.com/PrefectHQ/prefect/pull/15504

Comment (zzstoatzz) [reply src/prefect/settings.py]: updated to this in [ffea111](https://github.com/PrefectHQ/prefect/pull/15412/commits/ffea11121e50253408f7ccb5b21a8fc0457903ea)
```python
    deployment_concurrency_slot_wait_seconds: float = Field(
        default=30.0,
        ge=0.0,
        description=(
            "The number of seconds to wait before retrying when a deployment flow run"
            " cannot secure a concurrency slot from the server."
        ),
    )
```

**Review by cicdw — APPROVED** [review]

**Pull request merged**